src/DynamicNLPModels.jl

```diff
@@ -318,7 +364,125 @@ function LQDynamicModel(
     )
 end
 
+function _build_condensed_blocks(
+    s0, Q, R, A, B, N;
+    Qf = Q)
+  
+    ns = size(Q,1)
+    nu = size(R,1)
+  
+    # Define block matrices
+    block_B = zeros(ns*(N+1), nu*(N))
+    block_A = zeros(ns*(N+1), ns)
+    block_Q = SparseArrays.sparse([],[], Float64[], ns*(N+1), ns*(N+1))
+    block_R = SparseArrays.sparse([],[], Float64[], nu*(N), nu*(N))
+  
+    block_A[1:ns, 1:ns] .= Matrix(LinearAlgebra.I, ns, ns)
+  
+    # Define matrices for mul!
+    A_klast  = copy(A)
+    A_k      = similar(A)
+    AB_klast = similar(B)
+    AB_k     = similar(B)
+  
+    # Add diagonal of Bs and fill Q and R block matrices
+    for j in 1:(N)
+        row_range = (j*ns + 1):((j+1)*ns)
+        col_range = ((j-1)*nu+ 1):((j)*nu)
+        block_B[row_range, col_range] .= B
+  
+        block_Q[((j-1)*ns + 1):(j*ns), ((j-1)*ns + 1):(j*ns)] .= Q
+        block_R[((j-1)*nu + 1):(j*nu), ((j-1)*nu + 1):(j*nu)] .= R
+    end
+  
+    # Fill the A and B matrices
+    for i in 1:(N-1)
+        if i == 1
+            block_A[(ns+1):ns*2, :] .= A
+            LinearAlgebra.mul!(AB_k, A, B)
+            for k in 1:(N-i)
+                row_range = (1 + (k+1) * ns ):((k+2)*ns)
+                col_range = (1 + (k-1)*nu):((k)*nu)
+                block_B[row_range, col_range] .= AB_k
+            end
+            AB_klast = copy(AB_k)
+        else
+            LinearAlgebra.mul!(AB_k, A, AB_klast)
+            LinearAlgebra.mul!(A_k, A, A_klast)
+            block_A[(ns*i + 1):ns*(i+1),:] .= A_k
+  
+            for k in 1:(N-i)
+                row_range = (1 + (k+i) * ns):((k+i+1)*ns)
+                col_range = (1 + (k-1)*nu):((k)*nu)
+                block_B[row_range, col_range] .= AB_k
+            end
+  
+            AB_klast = copy(AB_k)
+            A_klast  = copy(A_k)
+        end
+    end
+  
+    LinearAlgebra.mul!(A_k, A, A_klast)
 
+    block_A[(ns*N +1):ns*(N+1), :] .= A_k
+    block_Q[((N)*ns + 1):((N+1)*ns), ((N)*ns + 1):((N+1)*ns)] .= Qf
+  
+    # build quadratic term
+    QB  = similar(block_B)
+    BQB = zeros(nu*N, nu*N)
+
+    LinearAlgebra.mul!(QB, block_Q, block_B)
+    LinearAlgebra.mul!(BQB, transpose(block_B), QB)
+    LinearAlgebra.axpy!(1, block_R, BQB)
+    
+    # build linear term 
+    h    = zeros(1, nu*(N))
+    s0TAT = zeros(1, size(block_A,1))
+    LinearAlgebra.mul!(s0TAT, transpose(s0), transpose(block_A))
+    LinearAlgebra.mul!(h, s0TAT, QB)
+  
+    # build constant term 
+    QAs = zeros(size(s0TAT,2), size(s0TAT,1))
+    h0  = zeros(1,1)
+    LinearAlgebra.mul!(QAs, block_Q, transpose(s0TAT))
+    LinearAlgebra.mul!(h0, s0TAT, QAs)
+  
+    c0 = h0[1,1]/2
+  
+  
+    return BQB, vec(h), c0, block_A, block_B, block_Q, block_R
```

Comment:
Instead of returning a undefined tuple, I would recommend returning either a named tuple 
```julia 
(Qc=BQB, hc=vec(h), c0=c0, ...)
```
or a struct `CondensedFormQP`